File: programmers/algorithm/level2/sung-silver/프로세스.java
import java.util.*;

class Solution {
    public class Process{
        int location;
        int priority;
        
        public Process(int location, int priority){
            this.location = location;
            this.priority = priority;
        }
        
        public int getLocation(){
            return location;
        }
        
        public int getPriority(){
            return priority;
        }
    
    }
    
    public int solution(int[] priorities, int location) {
        int answer = 0;
        Queue<Process> p = new LinkedList<>();
        List<Integer> sort_priorities = new ArrayList<>();
        
        for(int i=0; i<priorities.length ; i++){
            p.add(new Process(i, priorities[i]));
            sort_priorities.add(priorities[i]);
        }
        
        sort_priorities.sort(Comparator.reverseOrder());
        
        while(!p.isEmpty()){
            Process process = p.poll();
            int priority = sort_priorities.get(0);
            
            if(process.getPriority() == priority){
                answer++;
                if(process.getLocation() == location){
                    break;
                }
                sort_priorities.remove(0);
            }
            else{
                p.add(process);
            }
        }
        
        

        return answer;
    }
}

File: programmers/unanchoi/stack/프로세스.kt
import java.util.LinkedList

class Solution {
    fun solution(priorities: IntArray, location: Int): Int {
        var answer = 0
        val locationQueue = LinkedList<Int>()
        val queue:LinkedList<Int> = LinkedList()

        for (idx in priorities.indices) {
            queue.add(priorities[idx])
            locationQueue.add(idx)
        }

        var answerList = mutableListOf<Int>()

        while (queue.isNotEmpty()) {
            val num = queue.peek()
            val numIdx = locationQueue.peek()

            if (num == queue.maxOrNull()) {
                answerList.add(numIdx)
                queue.pop()
                locationQueue.pop()
            }
            else {
               queue.add(queue.pop())
                locationQueue.add(locationQueue.pop())
            }
        }

        println(answerList)

        for (idx in 0..answerList.size-1) {
            if (location == answerList[idx]) {
                return idx+1
            }
        }
        return 0
    }
}